Allow packetizer marshal / unmarshal to be consistent

boushley opened the pull request:
#### Description
Add a test to validate round tripping packets from Packetizer results in identical packets. In addition, initialize the CSRC so it matches in line with the suggestions in the issue comments.

#### Reference issue
Fixes #79

jech commented:
I'm not quite sure how empty slices are represented in Go, but does this cause one extra allocation per packet?

boushley commented:
That's a great question @jech, and I'm not sure of the answer. I believe the answer is that this doesn't result in an extra allocation though.

First I tested the `len(...)` and `cap(...)` of the empty slice `[]uint32{}` and their both `0`, so it's definitely not setting much aside if any.

Then I created a basic test:

```
package main

import (
	"fmt"
	"testing"
)

func TestEmptySlice(t *testing.T) {

	holder := [][]uint32{}

	for i := 0; i < 10_000_000; i++ {
		holder = append(holder, nil)
	}

	fmt.Printf("Complete, test count: %d\n", len(holder))
}
```

And used `go test -memprofile ... -bench .` to create a memory profile. I did this once for `nil` and once for `[]uint32{}` being appended to the holder slice.

The `alloc_space` data for the `nil` version looks like this:
```
Showing nodes accounting for 1.20GB, 99.43% of 1.21GB total
Dropped 16 nodes (cum <= 0.01GB)
      flat  flat%   sum%        cum   cum%
    1.20GB 99.43% 99.43%     1.20GB 99.43%  allocation-experiments.TestEmptySlice
         0     0% 99.43%     1.20GB 99.43%  testing.tRunner
```

And the `alloc_objects` data for the `nil` version looks like this:
```
Showing nodes accounting for 7182, 99.53% of 7216 total
Dropped 2 nodes (cum <= 36)
Showing top 5 nodes out of 15
      flat  flat%   sum%        cum   cum%
      5922 82.07% 82.07%       5922 82.07%  runtime.allocm
      1260 17.46% 99.53%       1260 17.46%  runtime.malg
         0     0% 99.53%       1822 25.25%  runtime.mcall
         0     0% 99.53%       4100 56.82%  runtime.mstart
         0     0% 99.53%       4100 56.82%  runtime.mstart0
```

The `alloc_space` data for the `[]uint32{}` version looks like this:
```
Showing nodes accounting for 1.20GB, 99.44% of 1.21GB total
Dropped 15 nodes (cum <= 0.01GB)
      flat  flat%   sum%        cum   cum%
    1.20GB 99.44% 99.44%     1.20GB 99.44%  allocation-experiments.TestEmptySlice
         0     0% 99.44%     1.20GB 99.44%  testing.tRunner
```

The `alloc_objects` data for the `[]uint32{}` version looks like this:
```
Showing nodes accounting for 7987, 99.59% of 8020 total
Dropped 2 nodes (cum <= 40)
Showing top 5 nodes out of 15
      flat  flat%   sum%        cum   cum%
      5466 68.15% 68.15%       5466 68.15%  runtime.allocm
      2521 31.43% 99.59%       2521 31.43%  runtime.malg
         0     0% 99.59%       1366 17.03%  runtime.mcall
         0     0% 99.59%       4100 51.12%  runtime.mstart
         0     0% 99.59%       4100 51.12%  runtime.mstart0
```

boushley commented:
To be sure that I would catch allocations should they happen I also ran a test where I was appending `[]uint32{1}`.

The `alloc_space` results for that are:
```
Showing nodes accounting for 1.24GB, 99.45% of 1.25GB total
Dropped 16 nodes (cum <= 0.01GB)
      flat  flat%   sum%        cum   cum%
    1.24GB 99.45% 99.45%     1.24GB 99.45%  allocation-experiments.TestEmptySlice
         0     0% 99.45%     1.24GB 99.45%  testing.tRunner
```

And the `alloc_objects` result for that is:
```
Showing nodes accounting for 2687057, 99.51% of 2700337 total
Dropped 16 nodes (cum <= 13501)
      flat  flat%   sum%        cum   cum%
   2687057 99.51% 99.51%    2687057 99.51%  allocation-experiments.TestEmptySlice
         0     0% 99.51%    2687057 99.51%  testing.tRunner
```

jech commented:
Great, thanks for your work.  The overhead, if any, is pretty minimal.
